src/main/scala/com/mozilla/telemetry/heka/package.scala

```diff
@@ -7,13 +7,63 @@
 package com.mozilla.telemetry
 
 import com.google.protobuf.ByteString
+import org.json4s._
+import org.json4s.jackson.JsonMethods._
+import org.json4s.JsonDSL._
+import scala.util.Try
 
 package object heka {
   class RichMessage(m: Message) {
+
+    /** Deserialize the Heka message as a native Scala Map. */
     def fieldsAsMap: Map[String, Any] = {
       val fields = m.fields
       Map(fields.map(_.name).zip(fields.map(field)): _*)
     }
+
+    /** Reconstructs a Heka message as a json4s JValue.
+      *
+      * The message is parsed and serialized into JSON. Keys that have been extracted
+      * into the toplevel document are inserted into to the proper place in the document.
+      *
+      * @return reconstructed JValue
+      */
+    def toJValue: JValue = {
+      val fields = Map(m.fields.map(_.name).zip(m.fields.map(fieldAsJValue)): _*)
+
+      lazy val submission = fields.getOrElse("submission", JNothing)
+      val payload = m.payload match {
+        case Some(payload: String) => {
+          val json = Try(parse(payload)).getOrElse(JNothing)
+          json match {
+            case JObject(x) => json
+            case _ => submission
+          }
+        }
+        case _ => submission
+      }
+
+      val (extractedKeys, metaKeys) =
+        fields
+          .keys
+          .filter(k => k != "submission")
+          .partition(k => k.contains("."))
```

Comment:
Instead of looking for a `.` in a field name to indicate that it should be parsed as JSON, we can look at the "representation" attribute of the field itself (as of [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1348337)). That has been deployed for nearly a year, so it should be available for pretty much everything in the S3 data store.

Reply:
Good point, but I'd like to defer this work since this doesn't affect correctness. The work downstream is pretty significant, like migrating the tests in jobs like MainSummaryView and Longitudinal. The representation attribute will need to be added to each of the fields in RichMessage fixtures.

Reply:
Sounds good, can you file a follow-up issue to take a look at that separately?